crates/core/tedge_agent/src/restart_manager/actor.rs

```diff
@@ -91,89 +112,100 @@ impl RestartManagerActor {
+                                command.with_error(error)
+                            }
+                        };
+
+                        Some(command)
+                    }
+                    StateStatus::Restart(RestartOperationStatus::Pending) => {
+                        let error = "The agent has been restarted but not the device";
+                        error!(error);
+                        Some(command.with_error(error.to_string()))
+                    }
+                    StateStatus::Software(_) | StateStatus::UnknownOperation => {
+                        error!("UnknownOperation in store.");
+                        None
+                    }
                 }
-            };
+            }
+            Err(err) => {
+                error!("Fail to read tedge-agent state: {err}");
+                None
+            }
+            Ok(_) => None,
         }
-        None
     }
 
-    async fn update_state_repository(
-        &mut self,
-        request: &RestartOperationRequest,
-    ) -> RestartOperationResponse {
-        let response = RestartOperationResponse::new(request);
+    async fn update_state_repository(&mut self, command: RestartCommand) -> RestartCommand {
         let state = State {
-            operation_id: Some(request.id.clone()),
+            operation_id: Some(command.cmd_id.clone()),
             operation: Some(StateStatus::Restart(RestartOperationStatus::Restarting)),
         };
 
         if let Err(err) = self.state_repository.store(&state).await {
-            error!(
+            let reason = format!(
                 "Fail to update the restart state in {} due to: {}",
                 self.state_repository.state_repo_path, err
             );
-            return response.with_status(OperationStatus::Failed);
+            error!(reason);
+            return command.with_error(reason);
         }
 
         if let Err(err) = create_tmp_restart_file(&self.config.tmp_dir).await {
-            error!(
+            let reason = format!(
                 "Fail to create a witness file in {} due to: {}",
                 self.config.tmp_dir, err
             );
-            return response.with_status(OperationStatus::Failed);
+            error!(reason);
+            return command.with_error(reason);
         }
 
-        response
+        command.with_status(CommandStatus::Executing)
     }
 
     async fn handle_restart_operation(&mut self) -> Result<(), RestartManagerError> {
-        let commands = self.get_restart_operation_commands().await?;
+        let commands = self.get_restart_operation_commands()?;
         for mut command in commands {
             match command.status().await {
                 Ok(status) => {
-                    if !status.success() {
-                        return Err(RestartManagerError::CommandFailed);
+                    if status.code().is_none() {
+                        return Err(RestartManagerError::CommandInterrupted {
```

Comment:
I don't think we should treat this as an error here as the script/process can be stopped via a signal as a result of the reboot handling (e.g. systemd killing the services quickly after `shutdown -r now` or `reboot` is executed in the command).

Instead we can just treat it like it was successful, and let the failed restart logic handle the case where no restart happens.